### PR TITLE
Rspec and Capybara/Selenium Setup

### PR DESCRIPTION
**Pull request by georgestam**
PR from issue #8 - Rspec and Capybara/Selenium Setup.

I have also written a couple of tests for the landing page and locale selector.

### CODE REVIEW

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  end
+  
+  config.after(:each) do |example|
+    assert_js_ok if example.metadata[:js] && !example.metadata[:skip_js_check]
+    DatabaseCleaner.clean
+    Warden.test_reset!
+  end
+  
+  config.around :each, :js do |ex|
+    ex.run_with_retry retry: (ci? ? 3 : 2), retry_wait: 3
```
Comment (georgestam) [inline spec/spec_helper.rb:80]: `ci?` to be modified as per above.

File: config/initializers/global_functions.rb [hunk 2]
```diff
@@ -20,4 +20,8 @@ def development_or_test?
 
 def production_or_staging?
   staging? || production?
+end
+
+def ci?
+  # [ENV['CODESHIP'], ENV['CI']].any? &:present?
```
Comment (georgestam) [inline config/initializers/global_functions.rb]: I assume that `ci` stands for Travis IC, which is not relevant here. I would like to remove this method, but I have 2 questions:
1. Not sure what is purpose of  line `80` from `spec_helper.rb` - ```ex.run_with_retry retry: (ci? ? 3 : 2), retry_wait: 3```. is it related with my `onerror`question above?
2. Where the variable `[ENV['CODESHIP']` is defined? If it is a key, I do not need any keys in my app to run the basic version of codeship. How can I tell to my app that I am in the codeship enviroment? Should not be enough to use `test?`(line 7 from `spec_helper.rb` uses `unless ENV['CODESHIP'])`

File: app/views/layouts/application.html.haml [hunk 3]
```diff
@@ -1,6 +1,13 @@
 !!!
 %html
   %head
+    - if test?
+      :javascript
+        window.errors = []
+        window.onerror = function(a, b, c){
+          window.errors.push([a, b, c]);
+          return false; // false: don't swallow exceptions.
+        }
```
Comment (georgestam) [inline app/views/layouts/application.html.haml:10]: Could you please remind me why this block of code is required? `onerror`seems an event to be triggered if an error occurs while loading an external file, but what are the `(a, b, c)` arguments?

**Review by georgestam — COMMENTED** [review]
@vemv, could you please review this PR? 

I have also added a few queries in code.

File: config/initializers/global_functions.rb (hunk 2 above)
Comment (vemv) [reply config/initializers/global_functions.rb]: CI means Continuous Integration. Not travis-specific

1. It configures the `rspec-retry` gem
2. Codeship sets an env var named `CODESHIP`. What `ci?` does is to determine whether code is being run by a CI provider. Either `ENV['CODESHIP']` or `ENV['CI']` would prove that

File: spec/spec_helper.rb (hunk 1 above)
Comment (vemv) [reply spec/spec_helper.rb:80]: No modification needed

File: app/views/layouts/application.html.haml (hunk 3 above)
Comment (vemv) [reply app/views/layouts/application.html.haml:10]: This backs the implementation of `assert_js_ok`. `assert_js_ok` must be run after every JS test. See kuina spec_helper and also `def assert_js_ok`

File: Gemfile [hunk 4]
```diff
@@ -6,7 +6,7 @@ gem 'figaro'
 gem 'jbuilder', '~> 2.0'
 gem 'pg'
 gem 'puma'
-gem 'rails', '5.0.0.1'
+gem 'rails', '5.0.1'
```
Comment (vemv) [inline Gemfile]: Generally, don't update `rails` lightly. Dedicate a commit or even a branch to it. When doing so, review why are you doing it, and also make sure to follow the rails upgrade guides (http://guides.rubyonrails.org/upgrading_ruby_on_rails.html)

This time is no big deal, next one remember it

File: Gemfile [hunk 5]
```diff
@@ -40,18 +40,38 @@ gem 'faker', '~> 1.7'
...
 group :development, :test do
   gem 'better_errors'
   gem 'binding_of_caller'
-  gem 'rubocop', '~> 0.48.1', require: false
 
+  gem "letter_opener"
   gem 'listen', '~> 3.0.5'
   gem 'pry-byebug'
   gem 'pry-rails'
+  gem 'rubocop', '~> 0.48.1'
```
Comment (vemv) [inline Gemfile]: * Keep Gemfile sorted alphabetically, leveraging atom `sort-lines` plugin
* Don't introduce random blank lines

File: spec/controllers/pages_spec.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+RSpec.describe PagesController, type: :controller do
+
+  describe 'landing page' do
+
+    context 'when it is loaded' do 
+      
+      it 'diplays the landing page' do
```
Comment (vemv) [inline spec/controllers/pages_spec.rb]: * This test doesn't really assert that it `diplays the landing page`. It merely asserts that the page `loads correctly`. Update the string so the message is truthful

* Try to no commit typos (`diplays`). Self-code-review habit might help

File: spec/features/locales_spec.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+describe 'locale' do
+
+  context 'when arabic is selected' do 
+    
+    it 'works' do
+      visit root_path(:en)
+      find("#change-to-arabic").click
```
Comment (vemv) [inline spec/features/locales_spec.rb]: Tests that assert that a page loads are too weak when you introduce interactions such as `find("#change-to-arabic").click`.

Instead (or additionally), assert that clicking the button does something.

Probably: `expect{ find("#change-to-arabic").click }.to change{ current_url }.from(x).to(y)`

**Review by vemv — CHANGES_REQUESTED** [review]

**Comment by georgestam** [comment]
@vemv I have implemented your requested changes.

File: app/views/shared/_navbar_top.html.haml [hunk 8]
```diff
@@ -137,9 +137,9 @@
           %a.cs-open-modal1
             %i.cs-in.fa.fa-globe>
             - if I18n.locale == :ea
-              = link_to (t '.english'), url_for( :locale => 'en')
+              = link_to (t '.english'), url_for( :locale => 'en'), id:"change-to-english"
```
Comment (vemv) [inline app/views/shared/_navbar_top.html.haml]: For the next time, space between `:` and `"`. Now leave it, let's move quick

File: spec/controllers/pages_spec.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+RSpec.describe PagesController, type: :controller do
+
+  describe 'landing page' do
+
+    context 'when it is loaded' do
```
Comment (vemv) [inline spec/controllers/pages_spec.rb:5]: `context` must never do nothing.

In this case it's doing nothing - it's a wrapper block which doesn't provide anything.

`context` blocks must always be accompanied of stuff like:

* `let` / `def`
* `before` / `after`

In this case, you can move `get 'index'` to a `before` block belonging to the `context` block. That way the `context` would do something - there would be a relationship between the text `'when it is loaded'` and an actual piece of code.

File: spec/features/locales_spec.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
+describe 'locale' do
```
Comment (vemv) [inline spec/features/locales_spec.rb]: You're not describing `locale`-  you're describing `locale switching` or something like that.

Pay attention to these

File: spec/features/locales_spec.rb [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  
+  before {
+    visit root_path(:en)
+  }
+
+  context 'when arabic is selected' do 
+    
+    it 'loads page' do
```
Comment (vemv) [inline spec/features/locales_spec.rb]: `'when arabic is selected it loads page'` does not read like a well-constructed, informative piece of English that reflects a feature of the app.

File: spec/features/locales_spec.rb [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+      find("#change-to-arabic").click
+      page_ok
+    end
+    
+    it 'changes language' do
```
Comment (vemv) [inline spec/features/locales_spec.rb]: Same comment as above for `when arabic is selected it changes language`

**Review by vemv — CHANGES_REQUESTED** [review]

**Comment by georgestam** [comment]
@vemv I have implemented the changes.

File: spec/features/locales_spec.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
+describe 'locale switching' do
+
+  context 'when arabic is selected' do
```
Comment (vemv) [inline spec/features/locales_spec.rb]: You say `when arabic is selected` but the code says `visit root_path(:en)`. You shouldn't lie! ;p

* move `visit root_path(:en)` at `describe` level
* Delete this test altogether:

```
     it 'loads page' do
       find("#change-to-arabic").click
       page_ok
     end
```
* Add a `page_ok` at the end of the remaining `it`
* Review the result and make sure it makes sense / adheres to guidelines

**Review by vemv — CHANGES_REQUESTED** [review]

File: spec/features/locales_spec.rb (hunk 13 above)
Comment (georgestam) [reply spec/features/locales_spec.rb]: Then, I will have to remove the `context` of this test as you mentioned before that it should always have some 'stuff' such as 'let / def' or 'before / after'. Correct?

**Comment by vemv** [comment]
Yes, probably

**Comment by georgestam** [comment]
@vemv Done.

**Review by vemv — APPROVED** [review]
Excellent, thank you!

I guess this was bit of a tiresome process, but better to   set a good level since the beginning

These will be less frequent, especially if you internalize these lessons.

**Comment by georgestam** [comment]
Thanks to your Victor!  💯

**Pull request merged**